### PR TITLE
Keep all maven deps in the central place for easier version management

### PR DESCRIPTION
**Pull request by liucijus**
I've been thinking how to make external maven deps management simpler and decided to move all maven deps into single location per Scala version. I hope this will bring better visibility on which deps we are using and help update them more routinely. 

My hope is that this refactoring helps towards:
- an easier manual or automated update of versions
- adding support of new Scala versions, eg #809
- a unified user experience when consuming rules with deps from repository macros
- less code duplication

**This is a breaking change for some repository users (eg. scrooge).** It's also not fully tested yet, marking PR as a draft and calling for your opinions.

### CODE REVIEW

**Review by blorente — COMMENTED** [review]
In general, this looks really good and a change in a direction I really love to see, thank you so much!

I've left a comment regarding some implementation details on the scrooge side. Don't really mind breaking changes, as long as we can keep the functionality described in some form.

Let me know if I can help any more!

File: twitter_scrooge/twitter_scrooge.bzl
Comment (liucijus) [reply twitter_scrooge/twitter_scrooge.bzl:33]: @blorente Do you want to keep using override approach, or would you consider moving dependencies to a toolchain? I have a branch for scrooge already, but I will make a PR after other toolchains PRs will progress: https://github.com/liucijus/rules_scala/tree/scrooge-toolchain

Comment (blorente) [reply twitter_scrooge/twitter_scrooge.bzl:33]: Moving to a toolchain would be great, as long as we keep avoiding the resolution for targets that we override.

We'd be happy to migrate to toolchains, looking forward to that PR, we can discuss the specifics there :)

I guess, if this PR lands before the scrooge toolchains PR, we would be blocked on upgrading `rules_scala` until the scrooge toolchains PR lands, but I think that is not a big problem :)

Comment (liucijus) [reply twitter_scrooge/twitter_scrooge.bzl:33]: I hope to get with the toolchain first, otherwise I will leave previous interface

**Comment by ittaiz** [comment]
build still fails

File: junit/junit.bzl
Comment (simuons) [inline junit/junit.bzl:8]: `maven_servers` is not passed to `repositories`

File: scala/scalafmt/scalafmt_repositories.bzl
Comment (simuons) [inline scala/scalafmt/scalafmt_repositories.bzl:48]: I think `maven_servers` should be taken from prams of `scalafmt_repositories `

File: junit/junit.bzl
Comment (liucijus) [reply junit/junit.bzl:8]: Thanks for catching, @simuons!

File: scala_proto/private/scala_proto_default_repositories.bzl
Comment (simuons) [inline scala_proto/private/scala_proto_default_repositories.bzl:14]: I think `major_version` is not passed to `repositories`

File: specs2/specs2.bzl
Comment (simuons) [inline specs2/specs2.bzl:17]: Maybe this `major_version = _extract_major_version(scala_version)` could be in `repositories` function? It would be possible to remove this duplicated line from other `*_repositories`

**Review by simuons — APPROVED** [review]
👍

File: specs2/specs2.bzl
Comment (or-shachar) [inline specs2/specs2.bzl:24]: If we're using the default servers all the time - why can't it be just the default value and skip the explicit parameter?

File: third_party/repositories/repositories.bzl
Comment (or-shachar) [inline third_party/repositories/repositories.bzl:22]: Oh I see you already call here for default value - so can we remove the explicit parameter from caller site?

File: third_party/repositories/repositories.bzl
Comment (or-shachar) [reply third_party/repositories/repositories.bzl:19]: I understand why you chose repositories... To keep current naming convention - but maybe it's an opportunity to name it better? `define_maven_external_dependencies` or something similar?

File: third_party/repositories/scala_2_11.bzl
Comment (or-shachar) [inline third_party/repositories/scala_2_11.bzl:1]: Awesome!! How did you create this list? Did you use some kind of generator?
What happens when the next dev wants to update something?

**Review by or-shachar — COMMENTED** [review]
Looks very good Vaidas!!  I think once this is merged we should prioritize creating a base list and have some sort of tool that will generate those files automatically for all scala flavors.

File: third_party/repositories/repositories.bzl
Comment (liucijus) [reply third_party/repositories/repositories.bzl:19]: I prefer shorter names, I think finding specific names with underscores is harder for users. I think name that includes `repositories` is right as it aligns with its only usecase repository bzl pattern. I don't expect any other usecases.

File: specs2/specs2.bzl
Comment (liucijus) [reply specs2/specs2.bzl:24]: Users may provide other server urls

File: third_party/repositories/repositories.bzl
Comment (liucijus) [reply third_party/repositories/repositories.bzl:22]: Users will need to be able to provide their own urls

File: third_party/repositories/scala_2_11.bzl
Comment (liucijus) [reply third_party/repositories/scala_2_11.bzl:1]: Currently it's a manual procedure to update deps

**Comment by liucijus** [comment]
@blorente, as while we have discussion on #1116, I have restored existing Twitter Scrooge repositories interface. I would appreciate if you could take a look.
Otherwise this PR is in ready to be merged state.

**Comment by blorente** [comment]
@liucijus Looks good to me, thanks for putting so much thought into scrooge!

**Review by ittaiz — APPROVED** [review]

**Pull request merged**
